panther_analysis_tool/main.py

```diff
@@ -434,6 +437,132 @@ def generate_hash(filename: str) -> bytes:
+        return 1, ""
+    release_url = (
+        f"https://api.github.com/repos/{args.github_owner}/{args.github_repository}/releases"
+    )
+    # setup appropriate https headers
+    headers = {"accept": "application/vnd.github.v3+json", "Authorization": f"token {api_token}"}
+    # check this tag doesn't already exist
+    response = requests.get(release_url + f"/tags/{args.github_tag}", headers=headers)
+    if response.status_code == 200:
+        logging.error("tag already exists %s", args.github_tag)
+        return 1, ""
+    # create the release directory
+    current_time = datetime.now().isoformat(timespec="seconds").replace(":", "-")
+    release_dir = args.out if args.out != "." else f"release-{current_time}"
+    # setup and generate release assets
+    return_code = setup_release(args, release_dir, api_token)
+    if return_code != 0:
+        return return_code, ""
+    # then publish to Github
+    return_code = publish_github(args.github_tag, args.body, headers, release_url, release_dir)
+    if return_code != 0:
+        return return_code, ""
+    logging.info("draft release (%s) created in repo (%s)", args.github_tag, release_url)
+    return 0, ""
+
+
+def clone_github(
+    owner: str, repo: str, branch: str, path: str, access_token: str
+) -> Tuple[int, str]:
+    repo_url = (
+        f"https://{access_token}@github.com/{owner}/{repo}"
+        if access_token
+        else f"https://github.com/{owner}/{repo}"
+    )
+    repo_dir = os.path.join(path, f"{repo}")
+    logging.info("Cloning %s branch of %s/%s", branch, owner, repo)
+    cmd = [
+        "git",
+        "clone",
+        "--branch",
+        branch,
+        "--depth",
+        "1",
+        "-c",
+        "advice.detachedHead=false",
+        repo_url,
+        repo_dir,
+    ]
+    result = subprocess.run(cmd, check=False, timeout=120)  # nosec
+    return result.returncode, ""
+
+
+def setup_release(args: argparse.Namespace, release_dir: str, token: str) -> int:
+    if not os.path.isdir(release_dir):
+        logging.info(
+            "Creating release directory: %s",
+            release_dir,
+        )
+        os.makedirs(release_dir)
+    # pull latest version of the github repo
+    return_code, _ = clone_github(
+        args.github_owner, args.github_repository, args.github_branch, release_dir, token
+    )
+    if return_code != 0:
+        return return_code
+    # generate zip based on latest version of repo
+    owd = os.getcwd()
+    # change dir to release directory
+    os.chdir(release_dir)
+    args.path = "."
+    # run generate assets from release directory
+    return_code, _ = generate_release_assets(args)
+    if return_code != 0:
+        return return_code
+    os.chdir(owd)
+    return 0
+
+
+def publish_github(tag: str, body: str, headers: dict, release_url: str, release_dir: str) -> int:
+    payload = {"tag_name": tag, "draft": True}
+    if body:
+        payload["body"] = json.dumps(body)
```

Comment:
What's the logic with the `json.dumps`? Just to escape everything?

Reply:
stringify the json, otherwise the github api complains about a malformed request.

Reply:
oh wait - this is the body - which is already a string. -_-  I'll remove